src/librustc/ty/layout.rs

```diff
@@ -1125,354 +835,760 @@ impl<'tcx> fmt::Display for LayoutError<'tcx> {
     }
 }
 
-impl<'a, 'tcx> Layout {
-    pub fn compute_uncached(tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                            param_env: ty::ParamEnv<'tcx>,
-                            ty: Ty<'tcx>)
-                            -> Result<&'tcx Layout, LayoutError<'tcx>> {
-        let success = |layout| Ok(tcx.intern_layout(layout));
-        let dl = &tcx.data_layout;
-        assert!(!ty.has_infer_types());
+#[derive(PartialEq, Eq, Hash, Debug)]
+pub struct CachedLayout {
```

Comment:
this is probably going to annoy you but I don't think `CachedLayout` is the *best* name (given that this type has nothing to do with being cached). I understand not wanting to name this "plain" `Layout` (because we want people to use `TyLayout`) but maybe `LayoutDetails` would be a better name?

Reply:
`CachedLayout` comes from contrast with `FullLayout`, it's not my first choice. As long as I don't rebase it into the middle of the commit list I don't have anything against renaming it.
`LayoutDetails` is a really great name!